packages/core/plugins/core-plugins-browser-internal/src/plugin_contract_resolver.ts

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { PublicMethodsOf } from '@kbn/utility-types';
+import type { PluginName } from '@kbn/core-base-common';
+import type {
+  PluginContractResolverResponse,
+  PluginContractMap,
+  PluginContractResolverResponseItem,
+} from '@kbn/core-plugins-contracts-browser';
+
+export type IRuntimePluginContractResolver = PublicMethodsOf<RuntimePluginContractResolver>;
+
+export class RuntimePluginContractResolver {
+  private dependencyMap?: Map<PluginName, Set<PluginName>>;
+  private setupContracts?: Map<PluginName, unknown>;
+  private startContracts?: Map<PluginName, unknown>;
```

Comment:
The actual "dynamic contract resolving" implementation.

Note that the browser and server-side implementations are dissociated, even if extremely similar. it's mostly that the types aren't the same (browser/server contract definitions) and trying to make the thing generic didn't feel like it was worth it.